### PR TITLE
Add 0017-missed-chain-events adr

### PR DESCRIPTION
**Pull request by bitwiseguy**
Add adr to document design and thought-process around restart-capable chain event tracking

### CODE REVIEW

File: .adr/0017-missed-chain-events.md
Comment (geoknee) [inline .adr/0017-missed-chain-events.md:9]:
```suggestion

It is critical that a go-nitro node can synchronise pertinent blockchain state, in order to successfully perform off chain protocols and protect against faulty challenges. For efficiency reasons, a node synchronises that state by listening to blockchain events. Missing those events can therefore cause the node's data to get out of sync with other nodes and the chain state. Events might be missed if the node has not been continuously running throughout the lifetime of any particular channel it is a participant in. For example, there may have been a period of downtime or loss of internet due to power loss or maintenance.

To address these concerns, a nitro node should check for any chain events it might have missed since it was last online, and queue those events to be processed as part of the node's initialization routines.
```

**Review by geoknee — COMMENTED** [review]
Forgot to click "submit" on this review!

File: .adr/0017-missed-chain-events.md
Comment (bitwiseguy) [reply .adr/0017-missed-chain-events.md:13]: Yes the `memstore` just checks from genesis every time when restarting

File: .adr/0017-missed-chain-events.md
Comment (bitwiseguy) [reply .adr/0017-missed-chain-events.md]: This bullet has now been removed and the design details section has been updated to explain how we are doing what you suggest: start listening for event logs first, then go back and get historical logs (using the `eventTracker` mutex to safely allow these things to happen simultaneously). https://github.com/statechannels/go-nitro/blob/missed-chain-events-adr/.adr/0017-missed-chain-events.md?plain=1#L73

File: .adr/0017-missed-chain-events.md
Comment (geoknee) [reply .adr/0017-missed-chain-events.md:13]: Ok, perhaps let us put that in the ADR then. Does the `engine` need to know about what kind of store it is dealing with? It only has an `interface` right?

Comment (bitwiseguy) [reply .adr/0017-missed-chain-events.md:13]: That's correct, the `engine` will just call the `store`'s interface, regardless of which `store` type is being used. During the `InitializeNode` function, we call `store.GetLastBlockNumSeen()`. For the `memstore` this will always return `0`. https://github.com/statechannels/go-nitro/blob/main/internal/node/node.go#L36

Comment (bitwiseguy) [reply .adr/0017-missed-chain-events.md]: Added sentence to explain `memstore` init: https://github.com/statechannels/go-nitro/pull/1624/commits/201969ba1cf3f96300a2affdfcd4bcd0b0ea2d61

Comment (geoknee) [reply .adr/0017-missed-chain-events.md]: Thanks!

File: .adr/0017-missed-chain-events.md
Comment (lalexgap) [inline .adr/0017-missed-chain-events.md:127]: > We could allow a way to forcibly process an old chain event,
I'm not sure if that would be enough to handle chain re-orgs, since we'd also have to handle a case where an event happens and then after the chain-org the event didn't happen. So we'd need a mechanism to detect an event no longer exists and "undo" it.

I think there may also be an inherit risk to chain re-orgs that we can really only solve by waiting? That's because chain events can trigger nitro nodes to exchanged signed states, and once these are sent out we can't really undo them. For example if Bob signs a state with outcome `[A:5,B:5]`  based on-chain channel balance of `10`, he's at risk if a chain re-org occurs and the on-chain channel balance  is actually `5` now.

File: .adr/0017-missed-chain-events.md
Comment (lalexgap) [inline .adr/0017-missed-chain-events.md:45]: Should this be numbered like the rest of the decisions?

File: .adr/0017-missed-chain-events.md
Comment (lalexgap) [inline .adr/0017-missed-chain-events.md:73]: >Also, an `ethereum.FilterQuery` cannot specify a `TxIndex` so it will be more computationally efficient to check the `TxIndex` in the `Channel` class rather than in the `chainservice` logic.

While the `ethereum.FilterQuery` doesn't support querying by `TxIndex` I'm not sure why this makes the chain service less computationally efficient? It seems unlikely that you'd be querying by a specific `TxIndex`.

File: .adr/0017-missed-chain-events.md
Comment (lalexgap) [inline .adr/0017-missed-chain-events.md:93]: I think this check might have to be tweaked a bit.

- `event.BlockNum() > c.LastChainUpdate.BlockNum` we're ignoring the case where `event.BlockNum==c.LastChainUpdate.BlockNum` which is where `TxIndex` comes into play
- We require the `TxIndex` always be larger then the last, even if the events are in different blocks.
I think we want something like:

```suggestion
	if event.BlockNum() > c.LastChainUpdate.BlockNum || (event.BlockNum() == c.LastChainUpdate.BlockNum) && event.TxIndex() > c.LastChainUpdate.TxIndex {
```

File: .adr/0017-missed-chain-events.md
Comment (bitwiseguy) [reply .adr/0017-missed-chain-events.md:29]: Yes I agree that the `memstore` should only be used in a test setting. I added the sentence about `memstore` based on @geoknee feedback in this comment: https://github.com/statechannels/go-nitro/pull/1624#discussion_r1312902801

File: .adr/0017-missed-chain-events.md
Comment (bitwiseguy) [reply .adr/0017-missed-chain-events.md:127]: Yes that makes sense about having to "undo" off-chain state changes. I am happy to remove the "Future considerations" section to make this adr cleaner. Would that be preferable or do you think this adr needs more thought and detail about how the node should handle chain re-orgs that defeat the `REQUIRED_BLOCK_CONFIRMATIONS` limit?

File: .adr/0017-missed-chain-events.md
Comment (bitwiseguy) [reply .adr/0017-missed-chain-events.md:73]: Would you be in favor of removing the one sentence you quoted above?

File: .adr/0017-missed-chain-events.md
Comment (bitwiseguy) [reply .adr/0017-missed-chain-events.md:93]: Great catch! I will work on a PR that implements the `Channel` class data protections mentioned above and link that here before merging this adr PR

File: .adr/0017-missed-chain-events.md
Comment (lalexgap) [reply .adr/0017-missed-chain-events.md:127]: I'm happy to remove this section (or even leave it) but I think further thought about chain re-orgs should probably be handled outside of this PR/ADR (due to the complexity).

File: .adr/0017-missed-chain-events.md
Comment (lalexgap) [reply .adr/0017-missed-chain-events.md:73]: I'd be happy to remove it or reword it? I just don't quite follow why checking the `TxIndex` in the `chainservice` is less efficient than if we do so in the channel.

File: .adr/0017-missed-chain-events.md
Comment (geoknee) [reply .adr/0017-missed-chain-events.md:29]: Agreed that `memstore` should be deprecated, since I think we can recover `memstore`-like functionality by configuring the `durablestore` a certain way anyway?

In my comment I was mostly concerned that the ADR was describing how the `durablestore` (implementation) should work, whereas it would be nice if it just described how a general `store` (interface) should work. I think not a huge deal either way.

Comment (bitwiseguy) [reply .adr/0017-missed-chain-events.md:29]: Would you guys prefer that I remove this sentence:
```
If the node uses a `memstore`, then the `lastBlockNumSeen` will always be set to `0` when the node is first initialized.
```
That is the only reference to `memstore` in this adr.

File: .adr/0017-missed-chain-events.md
Comment (bitwiseguy) [reply .adr/0017-missed-chain-events.md]: Removed that section: https://github.com/statechannels/go-nitro/pull/1624/commits/f69ab0eb34a0dcbc8c65c70ac2edfd051a682047

File: .adr/0017-missed-chain-events.md
Comment (bitwiseguy) [reply .adr/0017-missed-chain-events.md:96]: PR to make the above modifications to the `Channel` class: https://github.com/statechannels/go-nitro/pull/1659

**Review by lalexgap — APPROVED** [review]
🌋

File: .adr/0017-missed-chain-events.md
Comment (geoknee) [reply .adr/0017-missed-chain-events.md]: I don't feel strongly. Your call @bitwiseguy !

Comment (bitwiseguy) [reply .adr/0017-missed-chain-events.md]: Removed in https://github.com/statechannels/go-nitro/pull/1624/commits/f69ab0eb34a0dcbc8c65c70ac2edfd051a682047

File: .adr/0017-missed-chain-events.md
Comment (geoknee) [inline .adr/0017-missed-chain-events.md:30]:
```suggestion
3. By storing a `lastBlockNum` variable in the `store`,  the `chainservice` will not have to scan through as many old blocks when it first initializes and calls `checkForMissedEvents`. When the node is initialized it can read the `lastBlockNumSeen` from the `store`, and search for any chain events that occurred between that block and the current block. If the node uses a `memstore`, then the `lastBlockNumSeen` will always be set to `0` when the node is first initialized. When/how should the `lastBlockNumSeen` be updated?
```

File: .adr/0017-missed-chain-events.md
Comment (geoknee) [inline .adr/0017-missed-chain-events.md:116]: This sentence trails off?

Comment (bitwiseguy) [reply .adr/0017-missed-chain-events.md]: Fixed in https://github.com/statechannels/go-nitro/pull/1624/commits/06b800ff68d376a29ee776e21e9b8a074fcd8ff6

File: .adr/0017-missed-chain-events.md
Comment (bitwiseguy) [reply .adr/0017-missed-chain-events.md]: Change made in https://github.com/statechannels/go-nitro/pull/1624/commits/06b800ff68d376a29ee776e21e9b8a074fcd8ff6

**Pull request merged**
